modules/repository-datastax/src/main/java/com/intuit/wasabi/repository/cassandra/impl/AssignmentStats.java

```diff
@@ -0,0 +1,100 @@
+
+    public AssignmentStats() {
+        hourlyCountMap = new ConcurrentHashMap<>();
+        for (int hour = 0; hour <= 23; hour++) {
+            hourlyCountMap.put(hour, new ConcurrentHashMap<>());
+        }
+    }
+
+    public void incrementCount(Experiment experiment, Assignment assignment) {
+        int assignmentHour = getHour(assignment.getCreated());
+        Map<String, AtomicInteger> hourMap = hourlyCountMap.get(assignmentHour);
+        Experiment.ID id = experiment.getID();
+        Optional<Bucket.Label> labelOptional = Optional.ofNullable(assignment.getBucketLabel());
+        Bucket.Label bucketLabel = labelOptional.orElseGet(() -> NULL_LABEL);
+        AtomicInteger oldCount = hourMap.get(ExpBucket.getKey(id, bucketLabel));
+        if (oldCount == null) {
+            synchronized (lock) {
+                oldCount = hourMap.get(ExpBucket.getKey(id, bucketLabel));
+                if (oldCount == null) {
+                    AtomicInteger count = new AtomicInteger(1);
+                    hourMap.put(ExpBucket.getKey(id, bucketLabel), count);
+                } else {
+                    oldCount.getAndIncrement();
+                }
+            }
+        } else {
+            oldCount.getAndIncrement();
+        }
+    }
+
+    public int getCount(Experiment experiment, Bucket.Label bucketLabel, int assignmentHour) {
+        Map<String, AtomicInteger> hourMap = hourlyCountMap.get(assignmentHour);
+        return hourMap.get(ExpBucket.getKey(experiment.getID(), bucketLabel)).get();
+    }
+
+    public void writeCounts() {
+        Date completedHour = getLastCompletedHour(System.currentTimeMillis());
+        int assignmentHour = getHour(completedHour);
+        String day = getDayString(completedHour);
+        
+        Iterator it = hourlyCountMap.get(assignmentHour).entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry pair = (Map.Entry)it.next();
+            UUID experimentID = getExpUUID(pair);
+            String bucketLabel = getBucketLabel(pair);
+            int count = (int)pair.getValue();
+            for (int i = 0; i < count; i++){
+                hourlyBucketCountAccessor.incrementCountBy(experimentID, day, bucketLabel, assignmentHour);
+            }
+            System.out.println(pair.getKey() + " = " + pair.getValue());
+            it.remove(); // avoids a ConcurrentModificationException
+        }
+        hourlyCountMap.put(assignmentHour, new ConcurrentHashMap<>());
+    }
+
+    public Date getLastCompletedHour(long time) {
+        return new Date(time - 3600 * 1000);
+    }
+
+    public int getHour(Date completedHour) {
+        return Integer.parseInt(hourFormatter.format(completedHour));
+    }
+
+    public String getDayString(Date completedHour) {
+        return dayFormatter.format(completedHour);
+    }
+
+    public UUID getExpUUID(Map.Entry pair){
+        String expIDString = pair.getKey().toString().substring(0, UUID_LENGTH);
```

Comment:
@mooxter You might need to add error checking here to make sure the `pair` param is not null?